inst/include/conditionalSampler.h

```diff
@@ -0,0 +1,673 @@
+        dlogNCIt = CalcLogNC();
+        dlogNCPath += dlogNCIt;
+
+        //Normalise the weights.
+        pPopulation.SetLogWeight(pPopulation.GetLogWeight() - dlogNCIt);
+
+        //Check if the ESS is below some reasonable threshold and resample if necessary.
+        //A mechanism for setting this threshold is required.
+        double ESS = GetESS();
+        if(ESS < dResampleThreshold) {
+            nResampled = 1;
+            conditionalResample(rtResampleMode);
+        }
+        else {
+            nResampled = 0;
+            if(htHistoryMode == HistoryType::AL) {
+                uRSIndices = arma::linspace<arma::Col<unsigned int>>(0, N - 1, N);
+                //No resampling: set conditional index to previous iteration.
+                referenceTrajectoryIndices.at(T + 1) = referenceTrajectoryIndices.at(T);
+            }
+        }
+
+        //Normalise the weights
+        pPopulation.SetLogWeight(pPopulation.GetLogWeight() - CalcLogNC());
+
+        //Finally, the current particle set should be appended to the historical process.
+        if(htHistoryMode != HistoryType::NONE){
+            historyelement<Space> histel;
+            switch(htHistoryMode) {
+            case HistoryType::RAM:
+                histel.Set(N, pPopulation, historyflags(nResampled));
+                break;
+            case HistoryType::AL:
+                histel.Set(N, pPopulation, historyflags(nResampled), uRSIndices);
+                break;
+            //To avoid compiler warnings, HistoryType::NONE is handled
+            case HistoryType::NONE:
+                break;
+            }
+            History.push_back(histel);
+        }
+        //Increment the evolution time.
+        T++;
+
+        return ESS;
+    }
+    template <class Space, class Params>
+    void conditionalSampler<Space,Params>::Iterate(void)
+    {
+        IterateEss();
+        return;
+    }
+    template <class Space, class Params>
+    void conditionalSampler<Space,Params>::IterateUntil(long lTerminate)
+    {
+        while(GetTime() < lTerminate)
+        Iterate();
+    }
+    template <class Space, class Params>
+    void conditionalSampler<Space, Params>::conditionalResample(ResampleType::Enum lMode)
+    {
+        //Conditional resampling performed following the algorithms outlined in Appendix C of the paper "...".
+
+        //Initialize container for ancestor indices.
+        uRSIndices = arma::zeros<arma::Col<unsigned int> >(static_cast<int>(N));
+
+        switch(lMode) {
+        case ResampleType::MULTINOMIAL:
+        default:
+            {
+                //Algorithm 3
+                //Step 0:
+                //Sample conditional index K_t from appropriate version of the "lambda" distribution i.e. uniformly on {1,...,N} in case of Multinmial resampling.
+                // long Kt = floor(unif_rand()*static_cast<double>(N)); // sample lamba(k_{t}|w_{t-1}, k_{t-1})=1/N
+                long Kt = 0;
```

Comment:
What is done here doesn't seem to match the comment.

Reply:
Added comment.

 I changed this part in the last commit to better see if the conditional conditional trajectory is set correctly and also to save one draw. I'll leave 386 and 387 there for the moment and add an additional line above 388?

Reply:
Thanks; the main point was just to avoid a discrepancy between what was described and what was actually done... even to us that would be confusing in the end.